### DISCUSSION TITLE
direct_reparam

### DISCUSSION THREAD

**Discussion by zichenwang01** [mitsuba-renderer/mitsuba3#835] [Q&A] [answered]
Hi,

I used the built-in direct_reparam to run some experiments but could not get the correct result. 

My scene consists of a sphere and a plane at the back. The camera is at the front, and the point light is also at the front. 

<scene version="0.6.0">

    
    <integrator type="direct_reparam">
        <integer name="max_depth" value="2"/>
    </integrator>

    
    <sensor type="perspective">
        <float name="fov" value="45"/>

        
        <transform name="to_world">
            <lookat origin="0.5,0.5,1" target="0.5,0.5,0.5" up="0,1,0"/>
        </transform>

        
        <sampler type="independent">
            
            <integer name="sample_count" value="32"/>
        </sampler>

        
        <film type="hdrfilm">
            <integer name="width" value="512"/>
            <integer name="height" value="512"/>
            <rfilter type="gaussian">
                <float name="stddev" value="0.5"/>
            </rfilter>
            <boolean name="sample_border" value="true"/>
        </film>
    </sensor>

    
    <shape type="sphere">
        <float name="radius" value="0.1"/>
        <point name="center" x="0.5" y="0.5" z="0.5"/>
    </shape>

    <shape type="rectangle" id="Background">
        <transform name="to_world">
            <scale value="1.0"/>
            <lookat origin="0.5,0.5,0" target="0.5,0.5,0.5" up="0,1,0"/>
        </transform>
    </shape>

    
    
    <emitter type="point">
        <spectrum name="intensity" value="1"/>
        <point name="position" x="0.5" y="0.5" z="1"/>
    </emitter>

</scene>

Now I have a single parameter that moves the plane to its right and want to forward it to the rendered image. In theory, I should expect the image gradient to be zero everywhere, but instead I see a circle of gradients around the sphere. 

    """move right plane"""
    params = mi.traverse(scene)
    dr.enable_grad(params)

    p = mi.Float(1e-3)
    transform : mi.Transform4f = params['Background.to_world']
    transform.matrix[0,3] += p
    params['Background.to_world'] = transform
    params.update()

    image = mi.render(scene, params, spp=64)
    dr.enable_grad(image)
    dr.set_grad(p, 1)
    dr.forward_to(image)
    grad = dr.grad(image)

Am I misusing direct_reparam? Many thanks in advance.

**Comment by njroussel** [comment, marked as answer]
Hi @zichenwang01 

No, this is the correct utilization. I believe this is simply a limitation of the method ([project page](http://people.csail.mit.edu/sbangaru/projects/was-2020/)).
